sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSource.scala

```diff
@@ -330,4 +353,129 @@ object FileStreamSource {
 
     def size: Int = map.size()
   }
+
+  private[sql] class FileStreamSourceCleaner(
+      fileSystem: FileSystem,
+      sourcePath: Path,
+      baseArchivePathString: Option[String]) extends Logging {
+
+    private val sourceGlobFilters: Seq[GlobFilter] = buildSourceGlobFilters(sourcePath)
+
+    private val baseArchivePath: Option[Path] = baseArchivePathString.map(new Path(_))
+
+    def archive(entry: FileEntry): Unit = {
+      require(baseArchivePath.isDefined)
+
+      val curPath = new Path(new URI(entry.path))
+      val curPathUri = curPath.toUri
+
+      val newPath = buildArchiveFilePath(curPathUri)
+
+      if (isArchiveFileMatchedAgainstSourcePattern(newPath)) {
+        logWarning(s"Fail to move $curPath to $newPath - destination matches " +
+          s"to source path/pattern. Skip moving file.")
+      } else {
+        doArchive(curPath, newPath)
+      }
+    }
+
+    def delete(entry: FileEntry): Unit = {
+      val curPath = new Path(new URI(entry.path))
+      try {
+        logDebug(s"Removing completed file $curPath")
+
+        if (!fileSystem.delete(curPath, false)) {
+          logWarning(s"Fail to remove $curPath / skip removing file.")
+        }
+      } catch {
+        case NonFatal(e) =>
+          // Log to error but swallow exception to avoid process being stopped
+          logWarning(s"Fail to remove $curPath / skip removing file.", e)
+      }
+    }
+
+    private def buildSourceGlobFilters(sourcePath: Path): Seq[GlobFilter] = {
+      val filters = new scala.collection.mutable.MutableList[GlobFilter]()
+
+      var currentPath = sourcePath
+      while (!currentPath.isRoot) {
+        filters += new GlobFilter(currentPath.getName)
+        currentPath = currentPath.getParent
+      }
+
+      filters.toList
+    }
+
+    private def buildArchiveFilePath(pathUri: URI): Path = {
+      require(baseArchivePathString.isDefined)
+      val baseArchivePathStr = baseArchivePathString.get
+      val normalizedBaseArchiveDirPath = if (baseArchivePathStr.endsWith("/")) {
+        baseArchivePathStr.substring(0, baseArchivePathStr.length - 1)
+      } else {
+        baseArchivePathStr
+      }
+
+      new Path(normalizedBaseArchiveDirPath + pathUri.getPath)
+    }
+
+    private def isArchiveFileMatchedAgainstSourcePattern(archiveFile: Path): Boolean = {
```

Comment:
Can you add scaladoc explaining what this method does? I can't really figure out from the name, and the code is full of edge case checking. That begs for an explanation.

(My guess is that it's checking whether the archive dir is under the source directory? This looks way too complicated for a simple check like that.)

Reply:
> My guess is that it's checking whether the archive dir is under the source directory?

It checks whether the destination of archive "file" will be under the source path (which contains glob) to prevent the possibility of overwriting/re-reading as input. 

So actually the method name contains everything - if `SourcePattern` isn't a familiar representation, I'll change it to `SourcePath`. If we prefer to `MatchedWith` instead of `MatchedAgainst`, I'll change it. 

If we agree the method name explains what it is doing, we might be able to skip adding doc - especially it's a private method.

> This looks way too complicated for a simple check like that.

The method becomes complicated because of two reasons:

1) There's a tricky part in FileStreamSource: FileStreamSource doesn't only match the files which match the source path, but also matches the files which parent directory matches the source path. So we should consider both cases: 1) file itself is matched 2) parent directory is matched.

Please refer below comments in test code:

https://github.com/apache/spark/blob/bd8da3799dd160771ebb3ea55b7678b644248425/sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSourceSuite.scala#L736-L778

2) Checking with glob pattern is costly, so we would like to avoid the case via leveraging known information if possible. For example, when file is moved to archive directory, destination path will retain input file's path as suffix, so destination path can't be matched with source path if archive directory's depth is longer than 2. (Neither file nor parent directory of destination path can be matched with source path.)

Reply:
Given the size of your reply to my comment, do you still think that this shouldn't be explained in a comment at the top of the method itself?

Reply:
I didn't mean it shouldn't be explained in a scaladoc in method - I meant the method has comments in near code lines to explain. Sounds like the preference is adding scaladoc in this case; I'll leave a comment.